Wiki restructuring - The context (#1223)
Azure/ResourceModules: docs/wiki/The context - CARML library.md

> This page provides an overview of the CARML library. For further details refer to the dedicated Wiki section [The library](./The%20library)

Reviewer: ```suggestion
This page provides an overview of the CARML library. For an in-depth look, please refer to the dedicated Wiki section [The library](./The%20library)
```

- Version control: The applied configuration is version controlled and hence enables roll-backs & analysis.
- Repeatability: You can deploy your infrastructure in a repeatable fashion, hence minimizing the chance of manual errors.
- Reusability: You can reuse your automation to deploy the same infrastructure to different environments. For example leveraging a multi-stage deployment from a sandbox environment, via integration to production using the same code.

In the context of Bicep or ARM/JSON templates we usually leverage a combination of flexible templates that are deployed using multiple parameter files mapped to different scenarios.

# A module in CARML

In the context of _CARML_ we define a module as a reusable, template-based **building block** for Infrastructure as Code deployments of Azure resources.

Each module is generalized for maximum flexibility. Each template should be able to cover as many resource-specific scenarios as possible and not restrict the user by making assumptions on the users' behalf. Eventually, the injected parameters should decide what the template deploys.

> Furthermore, each module comes with default values for its optional parameters, a detailed documentation for its usage and one or multiple parameter files to proof its correctness.

Author: I wouldn't commit to that in the wiki since discussions are ongoing

> Furthermore, each module comes with default values for its optional parameters, a detailed documentation for its usage and one or multiple parameter files to proof its correctness.

Reviewer: Discussions on what we consider smart & secure yes - but I guess we already agreed that we should strive for that (whatever 'that' is). Hence the sentence has a `'should'` backdoor ;)

> Furthermore, each module comes with default values for its optional parameters, a detailed documentation for its usage and one or multiple parameter files to proof its correctness.

Author: still would add that once that is clarified and done

> Furthermore, each module comes with default values for its optional parameters, a detailed documentation for its usage and one or multiple parameter files to proof its correctness.

Reviewer: ok

## CARML module features

A CARML module should comply with the following characteristics:

- **Atomic unit**: Each module is tied to a specific resource type or strongly correlated services.
  > For example a virtual machine module also deploys related OS disks and network interfaces.

> - **Reusable**: Several modules can be combined together to create & orchestrate more complex Azure deployments like single services or workloads.

Reviewer: ```suggestion
- **Reusable**: Several modules can be combined together to create & orchestrate more complex architectures like single services or workloads.
```

  > For example resource group, network security group and virtual network modules can be combined to create a resource group hosting a virtual network and multiple subnets in it associated to specific NSGs.
- **Multi-purpose**: Each module aims to cover most of the main resource's capabilities, without the need to maintain multiple module instances for different use cases. Instead, the consumption of a generalized module happens through parameter files​.

>   > For example the same virtual machine module can deploy a Windows OS VM or a Linux based VM depending on input parameters.

Reviewer: ```suggestion
  > For example the same virtual machine module can deploy a Windows OS VM or a Linux-based VM depending on input parameters.
```

> - **Integrates child resources**: Each module can deploy one instance of a resource and, optionally, n-amount of its child-resources.

Reviewer: ```suggestion
- **Integrates child resources**: Each module can deploy one instance of a resource and, optionally, n instances of its child-resources.
```

>   > For example a key vault can optionally deploy n-amount of key vault access policies.

Reviewer: ```suggestion
  > For example, the key vault module can deploy _one_ key vault and optionally _n_ key vault access policies.
```
